### PR TITLE
Define multiple Solr cores

### PR DESCRIPTION
**Pull request by SeTeM**
The PR slightly changes a configuration DSL: `define_fields` block will be replaced by `define_cores`, so we'll be able to specify and use multiple Solr cores.

### Before
```ruby
Solr.configure do |config|
  config.define_fields do |f|
    f.field :manufacturer
    f.field :model
  end
end

request = Solr::Indexing::Request.new([doc])
request.run(commit: true)

request = Solr::Query::Request.new(search_term: 'term', fields: [:model])
request.run(page: 1, page_size: 10)

Solr.delete_by_id(3242343)
```

### After
```ruby
Solr.configure do |config|
  config.define_core(name: :listings) do |f|
    f.field :manufacturer
    f.field :model
  end

  config.define_core(name: :locations) do |f|
    f.field :country
    f.field :region
  end
end

request = Solr::Indexing::Request.new(core_name: :listings, documents: [doc])
request.run(commit: true)

request = Solr::Query::Request.new(core_name: :listings, search_term: 'term', fields: [:model])
request.run(page: 1, page_size: 10)

Solr.delete_by_id(3242343, core_name: :listings)
```

### CODE REVIEW

File: spec/query/request/boosting/numeric_field_value_match_boost_function_spec.rb [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
   describe '.to_solr_s' do
+    before do
+      Solr.configure do |config|
+        config.define_core(name: :'test-core') do |f|
+          f.field :machine_type
+        end
+      end
+    end
+
+    after(:each) do
```
Comment (drnluz) [inline spec/query/request/boosting/numeric_field_value_match_boost_function_spec.rb]: Isn't `:each` the default scope? I think it would be better to remove `(:each)` from here or add `(:each)` on `before` just to keep it consistent.

**Comment by Valve** [comment]
A high-level comment about this pull request:

I don't like how it forces things on you. Many users have a single core in their solr and just specify it in URL. 

It should be possible to just specify a URL with core and use solr w/out all these `core_name` everywhere, e.g. `http://localhost:8983/some-core`.

Secondly, many users (including us on Machinio) separate document types using an attribute, for example all documents are stored in a single core, but the type of the document can be set with 'type: Listing'.

So the library should accept an optional parameter of 'document_type' and set `type: document_type` internally (if it's available).

To summarize, there should be several modes of usage:

1. Core via URL and no document type (all documents are the same). In that case the API shouldn't require you to use neither `core_name` nor `document_type`.

2. [Most common use case]. Implicit core through the URL and explicit document type.  This should require only setting the document type on queries/deletes. 
We have to think about making it easy to do for users. 

3. Explicit core and explicit document type (both params are supplied).

Let me know what you think @SeTeM @drnluz

**Comment by Valve** [comment]
When working on this, please make the simplest options the default and only force to specify non-defaults when necessary.

Also I would allow to specify both cores/document types once in the configure block, then override the configuration in Request constructor and then again allow to override on a specific method level.

**Comment by Valve** [comment]
@SeTeM @drnluz  you can work together on this PR actually if you can parallelize it.

**Comment by SeTeM** [comment]
@Valve 
"Many users have a single core", could you provide a proof, please?
"many users separate document types using an attribute", could you provide a proof for this note too, please?

One core can contain several entity types: for example, listings and artificial listings. Both of them have the same fields and are used in one place. The most common example from guides is an entertainment site's search: it has one core with multiple entities (books, movies, TV shows) and they have almost the same defined fields.

Storing two completely different entities in one place isn't the best idea: e.g. a main search and an autocomplete. Different cores can both store documents with the same document type: you can search for books for a main search, or in an admin search, or in an autocomplete.

I've implemented the simplest API that @drnluz and I found.

I'm going to use a default core if there is only one defined core in a configuration.
For now we don't need to specify different URLs for cores, so passing a name to a configuration is enough.

> Also I would allow to specify both cores/document types once in the configure block, then override the configuration in Request constructor and then again allow to override on a specific method level.

Sorry, I don't get that.
Could you provide an exmaple of a DSL for a configuration file and an API for indexing/querying, please?

**Comment by SeTeM** [comment]
@Valve, @drnluz, actually, I think that `Explicit is better than implicit` (Zen of Python, PEP-20), so passing a name of a core is a better idea than using some default value.

What do you think about changing `core_name` to `core`?
```ruby
request = Solr::Indexing::Request.new(core: :listings, documents: [doc])
```
and so on.

**Comment by Valve** [comment]
> "Many users have a single core", could you provide a proof, please?
parts.machinio.com
machinio.com (before international sites appeared).
That's 2 use cased for you.
I also think that having a single core is the default in solr and accepting a default is always the most popular option. It would seem highly unlikely to me that most solr users would go into pains of creating and configuring multiple cores. 

> "many users separate document types using an attribute" could you provide a proof for this note too, please?
If you look at solr libraries for Ruby, the most pdopular one is sunspot. 
In sunspot that's exactly how it's done. See below:

```
[4] pry(main)> Lead.search do 
[4] pry(main)*   fulltext 'hello'
[4] pry(main)* end  
D, [2018-08-20T10:53:04.259637 #86386] DEBUG -- :   SOLR Request (307.8ms)  [ path=select parameters={fq: ["type:Lead"], q: "hello", fl: "* score", qf: "title_text customer_first_name_text customer_last_name_text customer_company_text customer_email_text customer_phone_text", defType: "edismax", start: 0, rows: 30} ]
```
**fq: ["type:Lead"]**

> Storing two completely different entities in one place isn't the best idea

Why? What's wrong with it?

> I'm going to use a default core if there is only one defined core in a configuration.

There is no such thing as "default core". You always must create a core. However the simplest scenario to access a core is to use SOLR_URL with a core name in it. 

> I've implemented the simplest API that @drnluz and I found.

yeah, and I'm suggesting an even simpler API for you :)

Examples of config override:

```
Solr.configure do 
   define_core :english
      define fields..
   end
end

# this request will go to :english core
req = Solr::Query::Request.new

# but it should be possible to override it
req = Solr::Query::Request.new(core_name: :spanish)
# now this one will go to :spanish (even though the configuration contains :english).

# now we have a reference to a request, we can override the core again
req.run(core_name: :french)
```

It should be possible to do the same with the document type.

I think that implementing an override on a run level is optional, however it's important to have it on a request constructor level.

This follows "cascade + override" approach (which you can use in CSS etc).

Let me know if you have any questions.

**Comment by SeTeM** [comment]
@Valve, machinio.com has two cores: `en` and `admin`.

**Comment by Valve** [comment]
@SeTeM  Machinio has 10 cores actually. But it used to have 1.

**Comment by Valve** [comment]
@SeTeM  I think it's fine to have `configure_core` call as a required call in configuration, but it shouldn't be a required argument for request methods.

**Comment by SeTeM** [comment]
@Valve, yes, it has 10 cores: 9 languages (machinio.com, machinio.de, machinio.fr and so on) and 1 admin core (it's used only on machinio.com).
So actually, machinio.com uses two cores simultaneously.
Anyway, you've provided 2 projects: the first of them already use multiple cores (machinio.com), and the other will have multiple cores soon (a main search and an autocomplete, that was a reason for making these changes in Solrb).

**Comment by Valve** [comment]
@SeTeM why parts will have multiple cores soon?

**Comment by SeTeM** [comment]
@Valve, different documents have different fields: listings have ~20 fields, but an autocomplete have 5 completely different fields. Why should they be in one core? They have nothing in common.

**Comment by Valve** [comment]
@SeTeM you should think about it like that:

it should be possible to install this gem and specify SOLR_URL in order to get started. No other actions should be necessary. 

Only when you need to change the default, add some additional field configs etc, only then you need to add configuration. 

With this change I need to specify a core name in multiple places before I can do anything useful. That's why I thought we should provide a better way for default solr installations

**Comment by drnluz** [comment]
On a previous project I worked using elasticsearch we used one index (equivalent of a solr core) for each data type we wanted to index. Some other people I know that work with ES use this approach as well.
Using one core or multiple cores depend on the project and the decision made by the developers. That's why I think it's a good idea to have both types of operation in this gem: Multiple core setup and single core setup.
If the user don't specify the core name on config just use the URL.

**Comment by Valve** [comment]
@drnluz yeah, I also think that way. I can install the gem, and run my app with a SOLR_URL and that should be enough to work with it. 

With this change, I'll need to write a configure block and pass `core_name: abc` into every request I'm instantiation. This creates too much friction and forces the library opinion on you.

If a user of this library does want to pass core names into configuration and requests, it should be possible, but it shouldn't be the only way. That's why I mentioned in my first comment:

> It should be possible to just specify a URL with core and use solr w/out all these core_name everywhere, e.g. http://localhost:8983/some-core.

And also it's true that it's possible to use separate cores per document types, but we cannot refuse that some companies/developers don't do it and separate using attributes.

**Comment by SeTeM** [comment]
@drnluz, @Valve, you tell something like "should be an option to use Postgres with only one table specified in a DB_URL".
Anyway, I got your idea and will add an ability to specify a core name only in `SOLR_URL`.

**Comment by Valve** [comment]
@SeTeM  you're over-reacting. Having multiple tables in a database is far more common than having multiple indexes in solr.

Also table names are in SQL standard: `select * from <table_name>`, there is no such equivalent in Solr.  

After all this arguing, you will take your nice core_name and push into in the SOLR_URL anyway, will you not? What's the point of arguing then?

**Comment by Valve** [comment]
@SeTeM 

> Anyway, I got your idea and will add an ability to specify a core name only in SOLR_URL.

you don't need to do it. It's already built into solr. All I'm asking of you is not to remove that ability :)

**Comment by SeTeM** [comment]
@Valve 

> Also table names are in SQL standard: select * from <table_name>, there is no such equivalent in Solr.

Actually, there is an equivalent: you can't request Solr without explicit core name in an URL: `http://localhost:8983/some-core/select`.

**Comment by SeTeM** [comment]
@Valve, the PR is ready.

There are 3 variants to specify solr's url:
```ruby
# variant 1, single-core
ENV['SOLR_URL'] = 'http://localhost:8983/solr/test-core' # URL with core's name

# variant 2, single-core
Solr.configure do |config|
  config.url = 'http://localhost:8983/solr/test-core' # URL with core's name

  config.define_core do |core|
    core.field :title
  end
end

# variant 3, single-core or multi-core
Solr.configure do |config|
  config.url = 'http://localhost:8983/solr' # URL without core's name

  config.define_core(name: 'test-core') do |core|
    core.field :title
  end

  config.define_core(name: 'test-core-2') do |core|
    core.field :model
  end
end
```

Specifying core's name to any methods is required only for multi-core configuration:
```ruby
# single-core
Solr::Indexing::Request.new(documents: [doc])

# multi-core
Solr::Indexing::Request.new(core_name: :listings, documents: [doc])
Solr::Indexing::Request.new(core_name: :cars, documents: [doc])
Solr::Indexing::Request.new(documents: [doc]) # Raises an error because core_name isn't specified
```

File: lib/solr/indexing/request.rb [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
-        @docs = docs
+      def initialize(core_name: nil, documents:)
+        @core_name = core_name
+        @documents = documents
       end
 
       def run(commit: false)
         # need to think how to move out commit data from the connection, it doesn't belong there
-        raw_response = connection(PATH, commit: commit).post_as_json(docs)
+        data = documents.map { |d| d.as_json(core_name: core_name || Solr.configuration.default_core_name) }
```
Comment (drnluz) [inline lib/solr/indexing/request.rb]: What about setting the default core name directly on initialize?
```
def initialize(core_name: Solr.configuration.default_core_name, documents:)
  @core_name = core_name
  @documents = documents
end 
```

File: README.md [hunk 3]
```diff
@@ -37,6 +38,26 @@ Solr.configure do |config|
...
+  # Define multiple cores with fields
+  # Core's URL is 'http://localhost:8983/cars'
+  config.define_core(name: :cars) do |f|
+    f.field :manufacturer
+    f.field :model
+  end
+end
+```
+
+Single core configuration:
```
Comment (Valve) [inline README.md]: can you please add to readme from simple to complex, starting from env var only and then single core and then multiple cores?

File: lib/solr.rb [hunk 4]
```diff
@@ -24,12 +24,12 @@ def configure
       yield configuration
     end
 
-    def delete_by_id(id, commit: false)
-      Solr::Delete::Request.new(id: id).run(commit: commit)
+    def delete_by_id(id, commit: false, core_name: Solr.configuration.default_core_name)
```
Comment (Valve) [inline lib/solr.rb]: I think it should be uniform everywhere, e.g. `delete_by_id(core: :listing)` and `Solr.configuration.default_core` for all user facing code.

File: lib/solr/configuration.rb [hunk 5]
```diff
@@ -1,32 +1,64 @@
...
       @faraday_options = { request: { timeout: 2, open_timeout: 8 } }
-      @url = ENV['SOLR_URL']
-      @fields = {}
+      @cores = {}
     end
 
-    def uri
-      @uri ||= Addressable::URI.parse(url)
+    def uri(core_name: nil)
+      Addressable::URI.parse(url(core_name: core_name))
```
Comment (Valve) [inline lib/solr/configuration.rb]: why did you decide to remove var caching? URI parsing is expensive and calling it for every request is undesirable

File: lib/solr/configuration.rb [hunk 6]
```diff
@@ -1,32 +1,64 @@
...
     end
 
-    def url
-      raise Errors::SolrUrlNotDefinedError unless @url
-      @url
+    def url=(value)
+      if value.nil?
+        raise Errors::SolrUrlNotDefinedError
```
Comment (Valve) [inline lib/solr/configuration.rb]: this should probably be argument error of some kind

File: lib/solr/configuration.rb [hunk 7]
```diff
@@ -1,32 +1,64 @@
...
+      raise Errors::SolrUrlNotDefinedError unless core_url
+      core_url
+    end
+
+    def default_core_name
+      default_core.name
+    end
+
+    def default_core
+      raise Errors::MultipleSolrCoresError if cores.count > 1
```
Comment (Valve) [inline lib/solr/configuration.rb]: it's fine to have multiple cores, but in this situation it's ambiguous

File: lib/solr/configuration.rb [hunk 8]
```diff
@@ -1,32 +1,64 @@
...
+        raise Errors::SolrUrlNotDefinedError
+      else
+        @url = value
+      end
     end
 
-    def define_fields
-      builder = Solr::FieldConfiguration::FieldDefinitionBuilder.new
+    def url(core_name: nil)
+      core_url = cores[core_name.to_sym]&.url if core_name
```
Comment (Valve) [inline lib/solr/configuration.rb]: shouldn't this entire method be cached in a hash keyed by core_name?

File: lib/solr/configuration.rb [hunk 9]
```diff
@@ -1,32 +1,64 @@
...
-      fields.merge!(builder.build)
+      core = builder.build
+      if cores.has_key?(core.name)
+        raise "A core with name '#{core.name}' has been already defined"
+      else
+        cores[core.name] = core
+      end
+    end
+
+    def null_core
```
Comment (Valve) [inline lib/solr/configuration.rb]: is there such thing as a "Null Core"?

File: lib/solr/core_configuration/core.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
+module Solr
+  module CoreConfiguration
+    class Core
```
Comment (Valve) [inline lib/solr/core_configuration/core.rb]: maybe for simpliciity we'll use `core` for user-facing APIs (method args etc) and for internal configuration use `CoreConfig` to distinguish it from the stringy core that's actually is a core_name?

File: lib/solr/core_configuration/core.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      attr_reader :name, :url, :fields
+
+      def initialize(name:, url:, fields:)
+        @name = name
```
Comment (Valve) [inline lib/solr/core_configuration/core.rb]: do we need both name and url? Isn't a core only defined by a name and then we can append it to the base URL?

File: lib/solr/core_configuration/core.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      end
+
+      private
+
+      def validate_options!
+        raise ArgumentError, 'pass name or url' if name.nil? && url.nil?
+      end
+    end
+
+    class NullCore < Core
```
Comment (Valve) [inline lib/solr/core_configuration/core.rb]: what's a Null Core?

File: lib/solr/delete/request.rb [hunk 13]
```diff
@@ -1,27 +1,30 @@
 module Solr
   module Delete
     class Request
-      include Solr::ConnectionHelper
-      using Solr::HashExtensions
+      include Solr::Support::ConnectionHelper
+      using Solr::Support::HashExtensions
 
-      PATH = '/update'.freeze
+      SOLR_DELETE_PATH = '/update'.freeze
```
Comment (Valve) [inline lib/solr/delete/request.rb]: why do we need this constant rename?

File: lib/solr/delete/request.rb [hunk 14]
```diff
@@ -1,27 +1,30 @@
...
+        raw_response = connection(SOLR_DELETE_PATH, core_name: core_name, commit: commit).post_as_json(delete_command)
         Solr::Response.from_raw_response(raw_response)
       end
 
       private
 
       def validate_delete_options!(options)
         options = options.deep_symbolize_keys
-        id, query = options.values_at(:id, :query)
+        core_name, id, query = options.values_at(:core_name, :id, :query)
```
Comment (Valve) [inline lib/solr/delete/request.rb]: why are you getting core_name value and not doing anything with it?

File: lib/solr/errors/multiple_solr_cores_error.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+module Errors
+  class MultipleSolrCoresError < StandardError
```
Comment (Valve) [inline lib/solr/errors/multiple_solr_cores_error.rb]: why having multiple cores is an error?

**Comment by SeTeM** [comment]
> Is it possible to define a core like that:
(https://github.com/machinio/solrb/pull/12#issuecomment-415218793)
Yes, sure

File: lib/solr/configuration.rb (hunk 5 above)
Comment (SeTeM) [reply lib/solr/configuration.rb]: good catch, thanks

File: lib/solr/configuration.rb (hunk 7 above)
Comment (SeTeM) [reply lib/solr/configuration.rb]: `default_core` shouldn't be called if there are more than 1 core defined

File: lib/solr/configuration.rb (hunk 8 above)
Comment (SeTeM) [reply lib/solr/configuration.rb]: yes, I'll do it

File: lib/solr/configuration.rb (hunk 9 above)
Comment (SeTeM) [reply lib/solr/configuration.rb]: Will rename to `NotSpecifiedCore`

File: lib/solr/core_configuration/core.rb (hunk 10 above)
Comment (SeTeM) [reply lib/solr/core_configuration/core.rb]: I already suggested renaming `core_name` to `core`: https://github.com/machinio/solrb/pull/12#issuecomment-414370224
Will do that.

File: lib/solr/delete/request.rb (hunk 14 above)
Comment (SeTeM) [reply lib/solr/delete/request.rb]: I used it to raise an error, but removed that check later

File: lib/solr/errors/multiple_solr_cores_error.rb (hunk 15 above)
Comment (SeTeM) [reply lib/solr/errors/multiple_solr_cores_error.rb]: Renamed

**Comment by SeTeM** [comment]
@Valve, I don't agree that one core is the most common use case.
Also, default core can't be defined for multi-cores configuration because for me explicit core's name is better than passing fewer arguments.
Could you provide any gem with your concept of default core?

File: README.md (hunk 3 above)
Comment (SeTeM) [reply README.md]: will improve README

File: lib/solr/core_configuration/core.rb (hunk 11 above)
Comment (SeTeM) [reply lib/solr/core_configuration/core.rb]: A core can be initialized with name and/or URL. For example, when a developer uses just SOLR_URL, the core doesn't have a name:
```ruby
# core doesn't have its name because it's included in the URL
Solr.configure do |config|
  config.url = 'http://localhost:8983/solr/test-core' # URL with core's name
end
```

**Comment by drnluz** [comment]
@Valve @SeTeM This is a very interesting discussion.
I tried to find some information about single core vs multi core setup but I couldn't find any good evidence online about which one is the most common. I tend to think multi core is, but again I couldn't find anything to support this other than my previous experiences with elasticsearch.

About core usage: 
If the user has defined only one core by naming it on configuration or through URL it's safe to let the user use it without specifying the core name.
If the user has multiple cores I think that having a default core setup is a good feature. This can be an optional feature: if the user specifies the default core name on config block, just use it. Otherwise he will have to pass `core_name` to every call. This way we allow the user to decide what works best for his case.

How to pass the core_name is another problem. The block implementation that V suggested can make things easier for us but I'm not sure if it is the best from the user point of view. I prefer having the core name as an argument like Vlad did, but we need to find a way to simplify this internally to avoid passing this argument many times through the entire stack.

File: lib/solr/configuration.rb (hunk 7 above)
Comment (Valve) [reply lib/solr/configuration.rb]: Yes, your error name is incorrect. It should be AmbiguousCoreError because having multiple cores is not an error, having multiple cores and not selecting one is

File: lib/solr/core_configuration/core.rb (hunk 11 above)
Comment (Valve) [reply lib/solr/core_configuration/core.rb]: core name is `test-core` in this case. In both cases core is identified by a path segment.

**Comment by Valve** [comment]
@SeTeM  I don't have examples of such gems

File: lib/solr/core_configuration/core.rb (hunk 11 above)
Comment (SeTeM) [reply lib/solr/core_configuration/core.rb]: @Valve, I already tried to find solr's URL schema, but without luck :(
Do you know where I can read more about it? That would be very useful.
For example, I can imagine several URLs:
```
http://localhost:8983/solr/test-core
http://localhost:8983/test-core
http://localhost:8983/solr
```
How can I be sure if path is a core name or not?

**Comment by SeTeM** [comment]
@drnluz, thanks for your answer.
I can let user specify default core if they want:
```ruby
Solr.configure do |config|
  config.url = 'http://localhost:8983/solr' # URL with core's name

  config.define_core(name: :listings, default: true) do |core|
    core.field :title
  end

  config.define_core(name: :users) do |core|
    core.field :name
  end
end
```
By default, all cores will use `default: false`, but @Valve and other users who like this behaviour may specify it manually.

@Valve, from my point of view, this is solution should satisfy both of us, that do you think?

**Comment by Valve** [comment]
sounds good thanks

File: lib/solr/indexing/document.rb [hunk 16]
```diff
@@ -13,12 +13,11 @@ def add_field(name, value)
       end
 
       # TODO: refactor & optimize this
-      def to_json(_json_context)
-        solr_fields = fields.map do |k, v|
-          solr_field_name = solarize_field(k)
+      def as_json(core:)
```
Comment (Valve) [inline lib/solr/indexing/document.rb]: I forgot to ask yesterday, why is this change necessary? Would you be able to revert it by specifying a current core on a lower level of configuration?

File: lib/solr/indexing/request.rb [hunk 17]
```diff
@@ -1,20 +1,22 @@
...
       # TODO: potentially make handlers configurable and have them handle the path
       PATH = '/update'.freeze
 
-      attr_reader :docs
+      attr_reader :core, :documents
 
-      def initialize(docs)
-        @docs = docs
+      def initialize(core: nil, documents:)
+        @core = core || Solr.configuration.default_core
```
Comment (Valve) [inline lib/solr/indexing/request.rb]: I think this should be moved to a lower level

File: lib/solr/query/request/boosting/geodist_function.rb [hunk 18]
```diff
@@ -17,7 +17,7 @@ def initialize(field:, latitude:, longitude:)
             freeze
           end
 
-          def to_solr_s
+          def to_solr_s(core:)
```
Comment (Valve) [inline lib/solr/query/request/boosting/geodist_function.rb]: this is wrong too, core shouldn't be required to convert something to a solr syntax

File: lib/solr/configuration.rb (hunk 8 above)
Comment (Valve) [reply lib/solr/configuration.rb]: thanks

**Comment by SeTeM** [comment]
I'll add `Solr.configuration.current_core`, so passing of core's name won't be required.

**Comment by SeTeM** [comment]
@Valve, @drnluz 
I'm adding a block `Solr.with_core` like @Valve described in [this comment](https://github.com/machinio/solrb/pull/12#pullrequestreview-148709587).
Hope this solution will satisfy you, but I would prefer an existing variant.

**Comment by Valve** [comment]
@SeTeM didn't you say that you'd add `Solr.configuration.current_core` method?

**Comment by SeTeM** [comment]
@Valve, @drnluz 
As I already mentioned, I implemented ideas from [this comment](https://github.com/machinio/solrb/pull/12#pullrequestreview-148709587):
1) Allow specifying default core (https://github.com/machinio/solrb/pull/12#issuecomment-415388096)
2) Single-core configuration doesn't require any additional code changes or blocks.
```ruby
Solr.configure do |config|
  config.url = 'http://localhost:8983/core-name'
end

Solr::Indexing::Request.new(documents: [doc])
```
3) Multi-core configuration can use default core without any additional code changes. Otherwise a block `Solr.with_core` can be used:
```ruby
Solr.with_core(:'non-default-core') do
  Solr::Indexing::Request.new(documents: [doc])
end
```
4) URL can be specified using `SOLR_URL` env var or using a configuration block:
```ruby
# without core definition
Solr.configure do |config|
  config.url = 'http://localhost:8983/core-name'
end

# with core definition
Solr.configure do |config|
  config.url = 'http://localhost:8983/core-name'

  config.define_core do |f|
    f.field :description
  end
end
```
5) The simplest configuration method: set `SOLR_URL` env var and that's it. A configuration file isn't required.
6) Tried to avoid passing core's name though methods using `Solr.current_core_config`.
